### PR TITLE
Adds weekly benchmarks, tencent test run

### PR DESCRIPTION
**Pull request by krzysztofpaliga**
### Description

This PR adds the support for:
- running weekly circuit benchmarks on Tencent Cloud, triggered by Github Actions
- running on demand arbitrary test commands on Tencent Cloud, triggered manually by using Github Actions

### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- .github/workflows/weekly-bench.yml
- .github/workflows/cloud-run.yml
- .github/weeklyBenchScripts/*
- .github/cloudRun/*

### Rationale

Since both circuit benchmarks and circuit tests require a large amount of memory, it is best to have them running in the Cloud.

### How Has This Been Tested?

By running through multiple times.

### CODE REVIEW

**Comment by pbeza** [comment]
FWIW - that's a nitpick, but I can see copy-pasted `43.130.90.57` in multiple files. Supposedly, this is the IP address of the Tencent Cloud server we use for testing. Maybe it's worth having it defined only in one place? (e.g. using GitHub secrets? same as with `BENCH_RESULTS_PASS`). Just a suggestion though.

**Review by Brechtpd — APPROVED** [review]
Approved because the code looks good, but I have no idea how any of this works! So hopefully somebody else that is more knowledgeable about this can also take a look and approve.

File: .github/weeklyBenchScripts/07_execBench.sh [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+current_time=$(date +'%H:%M:%S')
+echo "Current time: $current_time"
+echo $current_time > ~/bench_begin
+export RUST_BACKTRACE=1
+echo "DEGREE=$k ~/.cargo/bin/cargo test --profile bench bench_${run_suffix} -p circuit-benchmarks --features benches  -- --nocapture > \"$target_dir/results/$logfile\" 2>&1"
+DEGREE=$k ~/.cargo/bin/cargo test --profile bench bench_${run_suffix} -p circuit-benchmarks --features benches  -- --nocapture > "$target_dir/../results/$logfile" 2>&1
+
+
+exit 0
```
Comment (mratsim) [inline .github/weeklyBenchScripts/07_execBench.sh:68]: 01 to 05 don't end in exit 0. Is there a rational for using it there and not in the other?

File: .github/weeklyBenchScripts/07_execBench.sh [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    "copy")
+        run_suffix="copy_circuit_prover"
+        ;;
+    *)
+        echo "No proper value"
+        exit 1
+        ;;
+esac
+
+cd $target_dir;
```
Comment (mratsim) [inline .github/weeklyBenchScripts/07_execBench.sh:55]: quote variable

File: .github/weeklyBenchScripts/07_execBench.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+mkdir ../results
+logfile=$_date--${circuit}_bench-$k.proverlog
```
Comment (mratsim) [inline .github/weeklyBenchScripts/07_execBench.sh:58]: quote variable

File: .github/weeklyBenchScripts/07_execBench.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+current_time=$(date +'%H:%M:%S')
```
Comment (mratsim) [inline .github/weeklyBenchScripts/07_execBench.sh:60]: I think we should log in UTC time using `date -u`

File: .github/weeklyBenchScripts/08_processResults.sh [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+# Get the latest temp directory in the Provers home directory
+prover_latest_dir=$(ssh -i ~/.ssh/bench.pem -o StrictHostKeyChecking=no ubuntu@$PROVER_IP "bash -s" <<EOF
+ls -td -- "\$HOME"/CI_Prover_Benches/* | head -1
+EOF
+)
+
+prover_target_dir="$prover_latest_dir/zkevm-circuits"
+prover_results_dir="$prover_latest_dir/results"
+echo $prover_target_dir
```
Comment (mratsim) [inline .github/weeklyBenchScripts/08_processResults.sh:20]: quote variable

File: .github/weeklyBenchScripts/08_processResults.sh [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+cd $prover_results_dir
+./sadf.sh \$BENCH_BEGIN
+EOF
+scp -i ~/.ssh/bench.pem -o StrictHostKeyChecking=no ubuntu@$PROVER_IP:$prover_results_dir/*.stats $trigger_results_dir/
+
+# Prepare for and run data processing and db persistence
+
+l=$(echo $label | tr -d '"')
+circuit=$(echo $l |  awk '{print $1}')
+time=$(date +%Y-%m-%d_%H-%M-%S)
```
Comment (mratsim) [inline .github/weeklyBenchScripts/08_processResults.sh:48]: UTC time

**Review by mratsim — APPROVED** [review]
Overall, excellent work, some nits:

1. I think the scripts should contain at least:

    `set -eo pipefail` for immediate abort and proper exit code

    See https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html

2. I think dates should be reported in UTC, will be too confusing to have the dates sometimes in China time, sometimes in Singapore time, sometimes in US time depending on the instance we use.

3. Quoting with `"$MyVariable"` everywhere (or `"${MyVariable}"`). I did not check if all variables used are well declared, this might catch unseen dependencies (i.e. `set -u`)

4. If Github disconnects from our instance, I think it kills the bench, is that something we explicitly want or do we use tmux to keep the session alive? Due to VM cost, it likely makes sense to kill the instance.

File: .github/cloudRun/github-action-trigger.sh [hunk 7]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+sshpass -p $BENCH_RESULTS_PASS ssh -o StrictHostKeyChecking=no ubuntu@43.130.90.57 "bash -s" -- "$GITHUB_RUN_ID" "$BRANCH_NAME" <cloud-tests-trigger.sh
```
[the remark this answers is not in the document]
Comment (krzysztofpaliga) [reply .github/cloudRun/github-action-trigger.sh:3]: This IP is not meant to change.

File: .github/weeklyBenchScripts/00_installRust.sh [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+# Install the latest version of Rust
+echo "Installing the latest version of Rust..."
+rustup install stable
+
+# Set the latest version as the default
+rustup default stable
+
+# Print the installed Rust version
+echo "Rust has been installed successfully. Current version:"
```
[the remark this answers is not in the document]
Comment (krzysztofpaliga) [reply .github/weeklyBenchScripts/00_installRust.sh:21]: It is ok. It serves as starting point.

**Pull request merged**